File: day6/day6.js
const fs = require('fs');

// Read lines to data
const lines = fs.readFileSync('./entries.txt', 'utf-8').split('\n');

const charSet = new Set();
let countAns = 0;

// Calculate part one
lines.forEach((line) => {
  if (line.length !== 0) {
    [...line].forEach((char) => {
      charSet.add(char);
    });
  } else {
    countAns += charSet.size;
    charSet.clear();
  }
});

console.log(`Part one: ${countAns}`);

// Calculate part two
countAns = 0;
const charMap = new Map();
let countLine = 0;

lines.forEach((line) => {
  if (line.length !== 0) {
    countLine += 1;
    [...line].forEach((char) => {
      if (charMap.has(char)) {
        const newValue = charMap.get(char) + 1;
        charMap.delete(char);
        charMap.set(char, newValue);
      } else {
        charMap.set(char, 1);
      }
    });
  } else {
    charMap.forEach((value, key) => {
      countAns += value === countLine ? 1 : 0;
    });
    countLine = 0;
    charMap.clear();
  }
});

console.log(`Part two: ${countAns}`);

File: day6/day6.py
"""
Advent of code: Day 6
    Part 1: Count the distinct answers have "yes" from anyone
    Part 2: Count the distinct answers have "yes" from everyone from groups
"""

with open("entries.txt", "r") as file:
    data = list(line.rstrip() for line in file)

# Part One
line_string = ""
count_ans = 0

for line in data:
    if len(line) != 0:
        line_string += line
    else:
        count_ans += len(set(line_string))
        line_string = ""

print("Part One: {}".format(count_ans))

# Part Two
count_ans = 0
char_dict = dict()
count_line = 0

for line in data:
    if len(line) != 0:
        count_line += 1
        for char in line:
            if char in char_dict.keys():
                char_dict[char] += 1
            else:
                char_dict[char] = 1
    else:
        for char, value in char_dict.items():
            if value == count_line:
                count_ans += 1
        count_line = 0
        char_dict.clear()

print("Part Two: {}".format(count_ans))
